### DISCUSSION TITLE
is there any way to make output slot on Data mining on script?

### DISCUSSION THREAD

**Discussion by Knam-Ansys** [ansys/pyoptislang#359] [Q&A] [answered]
## is there any way to script this work?

![image](https://github.com/user-attachments/assets/703171e0-63a8-401d-a7bf-70fb963ec0e1)

Hello I am Kyungmin Nam in Korea.

I tried to make Data mining node's result to be output result, but I couldn't find any way.

Also, they have str location that I could't try pyoptislang's register output slot by location function.
(They only have dict location arg type)
{'location': 'OFilteredBestDesigns', 'name': 'OFilteredBestDesigns', 'reference_value': None}

**Comment by rfahlberg** [comment]
Hi @Knam-Ansys, what optiSLang and pyoptislang version are you using?

**Reply by Knam-Ansys** [reply]
Hi @rfahlberg I am using 2024R2 Optislang and current version of pyoptislang.(downloaded about 2 weeks ago.)

FYI:
I am trying to make aedt's snesitivity wizard and optimization wizard by using pyoptislang

![image](https://github.com/user-attachments/assets/a5c9ef15-08fc-426f-b308-cc7356b84cb3)

Other slot location like aedt node: {'dir': {'enum': ['input', 'output'], 'value': 'output'}, 'name': 'CMNF.OSL_Opt1.dB(St(Comm2,Comm1))[].Freq[GHz]'}
DataMining's location: 'OFilteredBestDesigns <-str type

**Comment by rfahlberg** [comment]
@Knam-Ansys I fear, the location registration support for the DataMining node is somewhat poor still in 24R2. It has been improved in 25R1, so if possible switch to 25R1.

In  24R2, in order to register the output slot, you will need to provide the full set of arguments to the [register_location_as_output_slot](https://optislang.docs.pyansys.com/version/stable/api/_autosummary/ansys.optislang.core.nodes.IntegrationNode.register_location_as_output_slot.html) method, including slot name and reference value. The reference value may be difficult to provide, as the type and structure depend on the location type. Easiest way would be to register the output slot once in GUI, connect to the project (via pyoptislang) and [obtain the currently registered output slots](https://optislang.docs.pyansys.com/version/stable/api/_autosummary/ansys.optislang.core.nodes.IntegrationNode.get_registered_output_slots.html) for the DataMining node (including reference values). Then you know, what exactly to provide to the register_location_as_output_slot method.

**Reply by Knam-Ansys** [reply]
@rfahlberg Thank you for your answer. I will ask if the 25R1 would be okay to use in our project or not.

But before that, I tried getting registered output slot and below is one of I got.

 {'location': 'OFilteredBestDesigns', 'name': 'OFilteredBestDesigns', 'reference_value': None}

I tried all of things that I could try but everything didn't go well. (str type location and nothing in ref)

I am really thank you for your support.

**Comment by rfahlberg** [comment]
@Knam-Ansys I fear, this is not yet supported with 24R2 for the DataMining Actor. 

@stefanmarth @ansDArnold do you have any suggestions?

**Reply by Knam-Ansys** [reply]
Thank you for your help

**Comment by Knam-Ansys** [comment]
Hello, @rfahlberg 

Thanks to your previous help, I tested it with 25R1 Optislang and the latest pyOptislang (24.12.03).
However, I couldn't find a solution, just as with 24R2.
![image](https://github.com/user-attachments/assets/66edaa3f-7406-4f37-a835-e4c7522b7948)

![image](https://github.com/user-attachments/assets/43c0d73f-d013-40c9-aeeb-476654c0a86b)
I still see str type location so that couldn't make command.

Could you possibly provide a very simple example of creating an output slot for Data Mining, as shown in the screenshot I tested?
Even if it's not an example using AEDT, any example that demonstrates passing filtered results from Data Mining as an output would be greatly appreciated!

Below is a base code that I tried to make output slot.
```python
from ansys.optislang.core import Optislang
from ansys.optislang.core.tcp.nodes import TcpProxySolverNodeProxy
from ansys.optislang.core.tcp import server_commands as sc
from ansys.optislang.core.tcp.osl_server import TcpOslServer
from ansys.optislang.core.tcp import server_queries as queries


def setup_test(osl):
    root_system = osl.application.project.root_system
    server: TcpOslServer = root_system._osl_server

    filter_node: TcpProxySolverNodeProxy = root_system.create_node(type_='DataMining', name="Data Mining")


    append_designs_property = {
        'OValidatedMDBPath': [{
            'First': {'name': 'AppendDesignsToFile'},
            'Second': [{'design_container': []}, {'string': 'IDesigns'}, {'string': 'IPath'}]
        }]
    }
    server.send_command(
        sc.set_actor_property(
            actor_uid=filter_node.uid,
            name='DataMiningManager',
            value={"id_filter_list_map": append_designs_property}
        )
    )

    # Register output command
    # filter_node.register_location_as_output_slot(
    #     location='OValidatedMDBPath', # <-which requires dict type 
    #     name='OValidatedMDBPath',
    #     reference_value=None)


    filter_node_get_output_slots = filter_node.get_output_slots()
    filter_node_actor_registered_output_slots = server.send_command(
        queries.actor_registered_output_slots(uid=filter_node.uid))
    print("1")


if __name__ == "__main__":
    opf_path = r"C:\Custom\Optislang\251_test.opf"
    osl = Optislang(ini_timeout=30, loglevel="info")
    osl.application.new()
    try:
        setup_test(osl)

    except Exception as e:
        print(f"ERROR: {str(e)}")
        import traceback

        print(traceback.format_exc())
    finally:
        osl.application.save_as(file_path=r"C:\Custom\Optislang\251_test2.opf")
        osl.dispose()
```
Thank you.

**Reply by rfahlberg** [reply, marked as answer]
@Knam-Ansys With 25R1,you should be able to do the following:
```
# create the DataMining node as in your script, but the node type is IntegrationNode
filter_node: IntegrationNode = root_system.create_node(type_='DataMining', name="Data Mining")

# set the filters/properties as in your script
# ...

# Load/Initialize the DataMining node in order to be able to register the locations
filter_node.load()

# Register the location as output slot
# This should work exactly as in your script, but name and reference_value parameters are not needed
filter_node.register_location_as_output_slot(location='OValidatedMDBPath')
```

For the property editing you can use the explicit API function ([set_property](https://optislang.docs.pyansys.com/version/stable/api/_autosummary/ansys.optislang.core.nodes.Node.set_property.html)) as well

**Reply by Knam-Ansys** [reply]
@rfahlberg Thanks for your helpful solution! 
I tried this in v242 and confirmed it only works correctly in v251. 
Your code example solved my issue perfectly. 
The step-by-step guidance for working with the IntegrationNode and registering location outputs was exactly what I needed. 
Thanks to your help, I was able to proceed with my project smoothly. I sincerely appreciate all the assistance you've provided so far. 👍


![image](https://github.com/user-attachments/assets/3c2630cf-5617-4ea3-a10e-760e38179ad4)
